src/csharp/Microsoft.Spark/Utils/UdfSerDe.cs

```diff
@@ -54,21 +62,168 @@ internal sealed class UdfData
+                return base.GetHashCode();
+            }
+
+            public override bool Equals(object obj)
+            {
+                return (obj is UdfData udfData) &&
+                    Equals(udfData);
+            }
+
+            public bool Equals(UdfData other)
+            {
+                return (other != null) &&
+                    TypeData.Equals(other.TypeData) &&
+                    (MethodName == other.MethodName) &&
+                    TargetData.Equals(other.TargetData);
+            }
         }
 
         [Serializable]
         internal sealed class TargetData
         {
             public TypeData TypeData { get; set; }
             public FieldData[] Fields { get; set; }
+
+            public override int GetHashCode()
+            {
+                return base.GetHashCode();
+            }
+
+            public override bool Equals(object obj)
+            {
+                return (obj is TargetData targetData) &&
+                    Equals(targetData);
+            }
+
+            public bool Equals(TargetData other)
+            {
+                if ((other == null) ||
+                    !TypeData.Equals(other.TypeData) ||
+                    (Fields?.Length != other.Fields?.Length))
+                {
+                    return false;
+                }
+
+                if ((Fields == null) && (other.Fields == null))
+                {
+                    return true;
+                }
+
+                return Fields.SequenceEqual(other.Fields);
+            }
         }
 
         [Serializable]
         internal sealed class FieldData
         {
-            public TypeData TypeData { get; set; }
-            public string Name { get; set; }
-            public object Value { get; set; }
+            public FieldData(object target, FieldInfo field)
+            {
+                object value = field.GetValue(target);
+
+                TypeData = SerializeType(field.FieldType);
+                Name = field.Name;
+                ValueData = (value != null)
+                    ? new ValueData(value)
+                    : null;
+            }
+
+            public TypeData TypeData { get; private set; }
+            public string Name { get; private set; }
+            public ValueData ValueData { get; private set; }
+
+            public override int GetHashCode()
+            {
+                return base.GetHashCode();
+            }
+
+            public override bool Equals(object obj)
+            {
+                return (obj is FieldData fieldData) &&
+                    Equals(fieldData);
+            }
+
+            public bool Equals(FieldData other)
+            {
+                return (other != null) &&
+                    TypeData.Equals(other.TypeData) &&
+                    (Name == other.Name) &&
+                    (((ValueData == null) && (other.ValueData == null)) ||
+                    ((ValueData != null) && ValueData.Equals(other.ValueData)));
```

Comment:
don't you need a tab here?

Reply:
Was auto formatted.  Manually adding 4 space indent here.